packages/keyring/src/keyring.ts

```diff
@@ -0,0 +1,333 @@
+          newAddresses,
+          results,
+          gapPerGroup
+        )
+        gapPerGroup = gap
+        activeAddresses.push(...newActiveAddresses)
+      }
+    }
+
+    this.addresses = [...this.addresses, ...activeAddresses]
+
+    return activeAddresses.map(this._getNonSensitiveAddressData)
+  }
+
+  // PRIVATE METHODS
+
+  private _getAddress = (addressHash: AddressHash): SensitiveAddressData => {
+    const address = this.addresses.find(({ hash }) => hash === addressHash)
+
+    if (!address) throw new Error(`Keyring: Could not find address with hash ${addressHash}`)
+
+    return address
+  }
+
+  private _generateAndCacheAddress = (props: GenerateAddressProps): SensitiveAddressData => {
+    const cachedAddress = this.addresses.find(({ index }) => index === props.addressIndex)
+
+    if (cachedAddress) return cachedAddress
+
+    const address = this._generateAddress(props)
+
+    this.addresses.push(address)
+
+    return address
+  }
+
+  private _generateAddress = ({
+    group,
+    addressIndex = 0,
+    skipAddressIndexes = []
+  }: GenerateAddressProps): SensitiveAddressData => {
+    if (group !== undefined && (!Number.isInteger(group) || group < 0 || group >= TOTAL_NUMBER_OF_GROUPS))
+      throw new Error(`Keyring: Could not generate address in group ${group}, group is invalid`)
+
+    let nextAddressIndex = skipAddressIndexes.includes(addressIndex)
+      ? findNextAvailableAddressIndex(addressIndex, skipAddressIndexes)
+      : addressIndex
+    let newAddressData = this._deriveAddressAndKeys(nextAddressIndex)
+
+    while (group !== undefined && addressToGroup(newAddressData.hash, TOTAL_NUMBER_OF_GROUPS) !== group) {
+      nextAddressIndex = findNextAvailableAddressIndex(newAddressData.index, skipAddressIndexes)
+      newAddressData = this._deriveAddressAndKeys(nextAddressIndex)
+    }
+
+    return newAddressData
+  }
+
+  private _deriveAddressesInGroup = (group: number, amount: number, skipIndexes: number[]): SensitiveAddressData[] => {
+    const addresses = []
+    const skipAddressIndexes = Array.from(skipIndexes)
+
+    for (let j = 0; j < amount; j++) {
+      const newAddress = this._generateAddress({ group, skipAddressIndexes })
+      addresses.push(newAddress)
+      skipAddressIndexes.push(newAddress.index)
+    }
+
+    return addresses
+  }
+
+  private _getNonSensitiveAddressData = ({
+    hash,
+    index,
+    publicKey
+  }: SensitiveAddressData): NonSensitiveAddressData => ({ hash, index, publicKey })
+
+  private _initFromMnemonic = (mnemonic: Uint8Array | null, passphrase: string) => {
+    if (this.root) throw new Error('Keyring: Secret recovery phrase already provided')
+    if (!mnemonic) throw new Error('Keyring: Secret recovery phrase not provided')
+
+    const isValid = validateMnemonic(mnemonic, wordlist)
+    if (!isValid) throw new Error('Keyring: Invalid secret recovery phrase provided')
+
+    const seed = mnemonicToSeedSync(mnemonic, wordlist, passphrase)
```

Comment:
This is where the mnemonic gets leaked (it is turned into a string in the implementation of this function: https://github.com/MetaMask/scure-bip39/blob/v2.1.0/src/index.ts#L142-L177

I don't understand why on Metamask it doesn't. It uses the same code: https://github.com/MetaMask/eth-hd-keyring/blob/v7.0.1/index.js#L298